File: native/android/PropertyCross/src/com/propertycross/android/presenter/SearchResultsPresenter.java
package com.propertycross.android.presenter;

import java.util.List;

import com.propertycross.android.events.Callback;
import com.propertycross.android.events.PropertyEventArgs;
import com.propertycross.android.events.PropertySelectedEvent;
import com.propertycross.android.events.UIEvent;
import com.propertycross.android.model.Property;
import com.propertycross.android.model.PropertyDataSource;
import com.propertycross.android.model.PropertyDataSourceResult;
import com.propertycross.android.model.PropertyListingsResult;
import com.propertycross.android.model.PropertyLocationsResult;
import com.propertycross.android.presenter.searchitems.SearchItem;

public class SearchResultsPresenter {

	public interface View {
		void setSearchResults(int totalResult, int pageNumber, int totalPages,
				List<Property> properties, String searchLocation);

		void setLoadMoreVisible(boolean isVisible);

		void setIsLoading(boolean isLoading);

		void loadMoreClicked(Callback<UIEvent> callback);

		void propertySelected(Callback<PropertySelectedEvent> callback);
	}

	private View view;
	private INavigationService navigationService;
	private SearchItem searchItem;
	private int pageNumber = 1;
	private PropertyDataSource dataSource;
	private List<Property> properties;
	private int totalResult;
	private int totalPages;
	private PropertyFinderPersistentState state;

	public SearchResultsPresenter(INavigationService navigationService,
			PropertyFinderPersistentState state,
			PropertyListingsResult results, SearchItem searchItem,
			PropertyDataSource dataSource) {
		this.state = state;
		this.navigationService = navigationService;
		this.searchItem = searchItem;
		this.dataSource = dataSource;
		this.properties = results.getData();
		this.totalResult = results.getTotalResult();
		this.totalPages = results.getTotalPages();
	}

	public void setView(View v) {
		view = v;
		view.loadMoreClicked(new Callback<UIEvent>() {

			public void complete(UIEvent event) {
				pageNumber++;
				view.setIsLoading(true);
				searchItem.findProperties(dataSource, pageNumber,
						new Callback<PropertyDataSourceResult>() {

							@Override
							public void complete(PropertyDataSourceResult result) {
								
								if (result instanceof PropertyListingsResult)
						        {
						          properties.addAll(((PropertyListingsResult)result).getData());
						          view.setSearchResults(totalResult, pageNumber, totalPages, properties, searchItem.getDisplayText());
						          view.setLoadMoreVisible(pageNumber < totalPages);
						          view.setIsLoading(false);
						        }
							}

						}, new Callback<Exception>() {

							@Override
							public void complete(Exception e) {
								view.setIsLoading(false);
							}
						});
			}
		});

		view.propertySelected(new Callback<PropertySelectedEvent>() {

			public void complete(PropertySelectedEvent event) {
				PropertyPresenter presenter = new PropertyPresenter(state,
						((PropertyEventArgs) event.getArguments())
								.getProperty());
				navigationService.pushPresenter(presenter);
			}
		});

		view.setSearchResults(totalResult, pageNumber, totalPages, properties,
				searchItem.getDisplayText());
		view.setLoadMoreVisible(pageNumber < totalPages);
	}
}

File: xamarin/common/Presenter/SearchResultsPresenter.cs
using System;
using System.Collections.Generic;
using PropertyCross.Model;

namespace PropertyCross.Presenter
{
  /// <summary>
  /// A presenter which renders a collection of search results.
  /// </summary>
  public class SearchResultsPresenter
  {
    /// <summary>
    /// The interface this presenter requires from the assocaited view.
    /// </summary>
    public interface View
    {
      /// <summary>
      /// Supplies the search results and meta-data.
      /// </summary>
      void SetSearchResults(int totalResult, int pageNumber, int totalPages,
        List<Property> properties, string searchLocation);

      /// <summary>
      /// Sets whether a 'load more' button should be displayed
      /// </summary>
      void SetLoadMoreVisible(bool visible);

      /// <summary>
      /// Sets whether to display a loading indicator.
      /// </summary>
      bool IsLoading { set; }

      event EventHandler LoadMoreClicked;

      event EventHandler<PropertyEventArgs> PropertySelected;
    }

    private View _view;
    
    private INavigationService _navigationService;
    
    private SearchItemBase _searchItem;

    private int _pageNumber = 1;

    private PropertyDataSource _dataSource;

    private List<Property> _properties;

    private int _totalResult;

    private int _totalPages;

    private PropertyCrossPersistentState _state;

    public SearchResultsPresenter(INavigationService navigationService, PropertyCrossPersistentState state,
      PropertyListingsResult results, SearchItemBase searchItem, PropertyDataSource dataSource)
    {
      _state = state;
      _navigationService = navigationService;
      _searchItem = searchItem;
      _dataSource = dataSource;
      _properties = results.Data;
      _totalResult = results.TotalResult;
      _totalPages = results.TotalPages;
    }

    public void SetView(View view)
    {
      _view = view;
      _view.LoadMoreClicked += View_LoadMoreClicked;
      _view.PropertySelected += View_PropertySelected;

      view.SetSearchResults(_totalResult, _pageNumber, _totalPages, _properties, _searchItem.DisplayText);
      view.SetLoadMoreVisible(_pageNumber < _totalPages);
    }

    private void View_PropertySelected(object sender, PropertyEventArgs e)
    {
      var presenter = new PropertyPresenter(_state, e.Property);
      _navigationService.PushPresenter(presenter);
    }

    private void View_LoadMoreClicked(object sender, EventArgs e)
    {
      _pageNumber++;
      _view.IsLoading = true;

      _searchItem.FindProperties(_dataSource, _pageNumber, response =>
      {
        if (response is PropertyListingsResult)
        {
          _properties.AddRange(((PropertyListingsResult)response).Data);
          _view.SetSearchResults(_totalResult, _pageNumber, _totalPages, _properties, _searchItem.DisplayText);
          _view.SetLoadMoreVisible(_pageNumber < _totalPages);
          _view.IsLoading = false;
        }
      }, error =>
      {
        _view.IsLoading = false;
      });
    }
  }
}
